### PR TITLE
[Autoscaler] Sync Files before Starting Docker

### LINKED ISSUE #17228: [Docker] [Autoscaler] Failure to setup head node on recycled head
Thanks to @richardliaw for consultation, any mistakes mine.

### What is the problem?
Failure to setup head node when running "ray up" on recycled head: cannot create directory '/tmp/ray_tmp_mount/default/~'

### Reproduction (REQUIRED)
ray up test-aws.yaml

Details:
Details:
 I had run "ray up test-aws.yaml" (will attach yaml) several times in the past, but today it started failing consistently.
 Adding -vvv we see (abbreviated):
```
 ubuntu@ip-172-31-12-183:~$ ray up -vvv test-aws.yaml
 <snip>
   Running `docker run --rm --name ray_container -d -it -v /tmp/ray_tmp_mount/default//home/ray/.aws:/home/ray/.aws -v /tmp/ray_tmp_mount/default//home/ubuntu/ludwig/:/home/ubuntu/ludwig/ -v /tmp/ray_tmp_mount/default//home/ray/.kaggle:/home/ray/.kaggle -e LC_ALL=C.UTF-8 -e LANG=C.UTF-8 --shm-size='21627607203.84b' --runtime=nvidia --net=host ludwigai/ludwig-ray-gpu:master bash`
     Full command is `ssh -tt -i ~/.ssh/aws-ludwig.pem -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes -o ExitOnForwardFailure=yes -o ServerAliveInterval=5 -o ServerAliveCountMax=3 -o ControlMaster=auto -o ControlPath=/tmp/ray_ssh_1d41c853af/c21f969b5f/%C -o ControlPersist=10s -o ConnectTimeout=120s ubuntu@34.221.222.129 bash --login -c -i 'true && source ~/.bashrc && export OMP_NUM_THREADS=1 PYTHONWARNINGS=ignore && (docker run --rm --name ray_container -d -it -v /tmp/ray_tmp_mount/default//home/ray/.aws:/home/ray/.aws -v /tmp/ray_tmp_mount/default//home/ubuntu/ludwig/:/home/ubuntu/ludwig/ -v /tmp/ray_tmp_mount/default//home/ray/.kaggle:/home/ray/.kaggle -e LC_ALL=C.UTF-8 -e LANG=C.UTF-8 --shm-size='"'"'21627607203.84b'"'"' --runtime=nvidia --net=host ludwigai/ludwig-ray-gpu:master bash)'`
 9984013237ed478e3dd50c96a0137e31dc6564f2e0b1b47d187eb3902b0f0bb4
 Shared connection to 34.221.222.129 closed.
   Running `mkdir -p /tmp/ray_tmp_mount/default/~ && chown -R ubuntu /tmp/ray_tmp_mount/default/~`
     Full command is `ssh -tt -i ~/.ssh/aws-ludwig.pem -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes -o ExitOnForwardFailure=yes -o ServerAliveInterval=5 -o ServerAliveCountMax=3 -o ControlMaster=auto -o ControlPath=/tmp/ray_ssh_1d41c853af/c21f969b5f/%C -o ControlPersist=10s -o ConnectTimeout=120s ubuntu@34.221.222.129 bash --login -c -i 'true && source ~/.bashrc && export OMP_NUM_THREADS=1 PYTHONWARNINGS=ignore && (mkdir -p /tmp/ray_tmp_mount/default/~ && chown -R ubuntu /tmp/ray_tmp_mount/default/~)'`
 mkdir: cannot create directory '/tmp/ray_tmp_mount/default/~': Permission denied
 Shared connection to 34.221.222.129 closed.
   New status: update-failed
   !!!
   {'message': 'SSH command failed.'}
   SSH command failed.
   !!!  Failed to setup head node.
 ubuntu@ip-172-31-12-183:~$

My file mounts are:
 file_mounts:
     {
         /home/ray/.aws: /home/ubuntu/.aws,
         /home/ubuntu/ludwig/: /home/ubuntu/ludwig,
         /home/ray/.kaggle: /home/ubuntu/.kaggle,
     }
Looking at the ray head:
 ubuntu@ip-172-31-16-91:~$ ls -alR /tmp/ray_tmp_mount/
 /tmp/ray_tmp_mount/:
 total 36
 drwxr-xr-x  3 root root  4096 Jul 20 21:33 .
 drwxrwxrwt 10 root root 28672 Jul 20 21:43 ..
 drwxr-xr-x  3 root root  4096 Jul 20 21:33 default/tmp/ray_tmp_mount/default:
 total 12
 drwxr-xr-x 3 root root 4096 Jul 20 21:33 .
 drwxr-xr-x 3 root root 4096 Jul 20 21:33 ..
 drwxr-xr-x 4 root root 4096 Jul 20 21:33 home/tmp/ray_tmp_mount/default/home:
 total 16
 drwxr-xr-x 4 root root 4096 Jul 20 21:33 .
 drwxr-xr-x 3 root root 4096 Jul 20 21:33 ..
 drwxr-xr-x 4 root root 4096 Jul 20 21:33 ray
 drwxr-xr-x 3 root root 4096 Jul 20 21:33 ubuntu/tmp/ray_tmp_mount/default/home/ray:
 total 16
 drwxr-xr-x 4 root root 4096 Jul 20 21:33 .
 drwxr-xr-x 4 root root 4096 Jul 20 21:33 ..
 drwxr-xr-x 2 root root 4096 Jul 20 21:33 .aws
 drwxr-xr-x 2 root root 4096 Jul 20 21:33 .kaggle/tmp/ray_tmp_mount/default/home/ray/.aws:
 total 8
 drwxr-xr-x 2 root root 4096 Jul 20 21:33 .
 drwxr-xr-x 4 root root 4096 Jul 20 21:33 ../tmp/ray_tmp_mount/default/home/ray/.kaggle:
 total 8
 drwxr-xr-x 2 root root 4096 Jul 20 21:33 .
 drwxr-xr-x 4 root root 4096 Jul 20 21:33 ../tmp/ray_tmp_mount/default/home/ubuntu:
 total 12
 drwxr-xr-x 3 root root 4096 Jul 20 21:33 .
 drwxr-xr-x 4 root root 4096 Jul 20 21:33 ..
 drwxr-xr-x 2 root root 4096 Jul 20 21:33 ludwig/tmp/ray_tmp_mount/default/home/ubuntu/ludwig:
 total 8
 drwxr-xr-x 2 root root 4096 Jul 20 21:33 .
 drwxr-xr-x 3 root root 4096 Jul 20 21:33 ..
```
Deleting the /tmp/ray_tmp_mount directory from the cached VM did not make any difference.

Setting `cache_stopped_nodes: False` under `provider` in the cluster configuration made the problem go away.
In this case, the /tmp/ray_tmp_mount related code is not executed

### PR DESCRIPTION
**Pull request by ijrsvt**
### A few pieces of context:
* When a docker container is started with a [mount](https://docs.docker.com/storage/volumes/) where the `host_path` **does not currently exist**, it will be created with `root` ownership.
* When most Linux machines are stopped & started the `/tmp` directory is deleted

### What is the issue:
1. Node Reuse is Enabled
2. A cluster is started (with file mounts), and eventually shutdown.
3. The Ray cluster is restarted and the head node is pulled from the **cache**.
4. `run_init` is executed **before** file mounts are synced
	a.  This causes `/tmp/ray_tmp_mount/`, `/tmp/ray_tmp_mount/<cluster_name>/` and `/tmp/ray_tmp_mount/<cluster_name>/<mount_name_1>` are all created _by docker and with **root** ownership_.
    b. `rsync` happens, and tries to copy files to `/tmp/ray_tmp_mount/<cluster_name>/<mount_name_1>` and fails because `ubuntu` (the default user on the host) cannot edit `root`-owned files.

### The solution:
Flip `4b` and `4a` to ensure that rsync creates the `/tmp/ray_tmp_mount/<cluster_name>/<mount_name_1>` path

## Related issue number
Closes #17228

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by ijrsvt** [comment]
bump @DmitriGekhtman @mwtian

**Review by DmitriGekhtman — APPROVED** [review]
lgtm!
probably you've already verified the fix manually/experimentally. if not, that's of course good to do before merging

File: python/ray/autoscaler/_private/command_runner.py [hunk 1]
```diff
@@ -820,6 +821,12 @@ def run_init(self, *, as_head, file_mounts, sync_run_yet):
                     run_env="host")
 
         if (not container_running) or requires_re_init:
+            if not sync_run_yet:
+                # Do not start the actual image as we need to run file_sync
+                # first to ensure that all folders are created with the
+                # correct ownership. Docker will create the folders with
+                # `root` as the owner.
+                return True
```
Comment (mwtian) [inline python/ray/autoscaler/_private/command_runner.py:829]: I'm curious about the other code path of returning True in this function: why is re-initialization needed when docker_run_executed is True? Do we need to rerun some commands?

**Review by mwtian — APPROVED** [review]

File: python/ray/autoscaler/_private/command_runner.py (hunk 1 above)
Comment (ijrsvt) [reply python/ray/autoscaler/_private/command_runner.py:829]: The main reason is that the `return_value` is `False` when the cluster is already in a good state. The `True` signfies that the cluster is in a state that requires further initialization.

**Comment by ijrsvt** [comment]
@DmitriGekhtman I am including test output from more recently testing it:
```
# ray down 
Stopping instances i-01b5aa8323f3fca56 (to terminate instead, set `cache_stopped_nodes: False` under `provider` in the cluster configuration)
Terminating instances i-081b9e2e5ba573406 (cannot stop spot instances, only terminate)
Requested 2 nodes to shut down. [interval=1s]
0 nodes remaining after 5 second(s).
No nodes remaining.

(ray) ubuntu@ip-172-31-47-243:~/ray$ ray up myaws.yaml -y -vvvvvvvvvvv
Cluster: CLUSTERNAME
...

No head node found. Launching a new cluster. Confirm [y/N]: y [automatic, due to --yes]

Acquiring an up-to-date head node
  Reusing nodes i-01b5aa8323f3fca56. To disable reuse, set `cache_stopped_nodes: False` under `provider` in the cluster configuration.
  Stopping instances to reuse
    Waiting for instance i-01b5aa8323f3fca56 to stop
  Launched a new head node
  Fetching the new head node
  
....
Useful commands
  Monitor autoscaling with
    ray exec /home/ubuntu/ray/myaws.yaml 'tail -n 100 -f /tmp/ray/session_latest/logs/monitor*'
  Connect to a terminal on the cluster head:
    ray attach /home/ubuntu/ray/myaws.yaml
  Get a remote shell to the cluster manually:
    ssh -tt -o IdentitiesOnly=yes -i /home/ubuntu/.ssh/ray-autoscaler_9_us-west-1.pem ubuntu@13.56.252.103 docker exec -it ray_container /bin/bash
```

And the YAML has:
```
provider:
    type: aws
    region: us-west-1
    availability_zone: us-west-1a,us-west-1b
    cache_stopped_nodes: True # If not present, the default is True.

file_mounts:
    "/home/ray/abc" : "/home/ubuntu/ray/python/ray/serve"
docker:
    image: "rayproject/ray-ml:1.4.0-py38-gpu"
    container_name: "ray_container"
```

**Pull request merged**